sha2/src/lib.rs

```diff
@@ -13,23 +13,37 @@
 #[cfg(not(any(target_arch = "x86_64", target_arch = "x86", target_arch = "aarch64")))]
 compile_error!("crate can only be used on x86, x86-64 and aarch64 architectures");
 
+cpufeatures::new!(cpuid_avx2, "avx2");
+
 #[link(name = "sha256", kind = "static")]
+#[allow(dead_code)]
 extern "C" {
     fn sha256_compress(state: &mut [u32; 8], block: &[u8; 64]);
+    fn sha256_transform_rorx(state: &mut [u32; 8], block: *const [u8; 64], num_blocks: u64);
 }
 
 /// Safe wrapper around assembly implementation of SHA256 compression function
+///
 #[inline]
 pub fn compress256(state: &mut [u32; 8], blocks: &[[u8; 64]]) {
-    for block in blocks {
-        unsafe { sha256_compress(state, block) }
+    let token: cpuid_avx2::InitToken = cpuid_avx2::init();
+
+    if token.get() {
```

Comment:
BTW it's better to use `cpuid_avx2::get()` directly inside this branch without getting the token. The current code checks the underlying atomic value twice.